### ISSUE TITLE
Avoid querying ALL_JSON_COLUMNS on Oracle 12.1.0.1.0

### ISSUE THREAD

**Issue by ahalim7** [jOOQ/jOOQ#10494]
I used to generate Jooq over `Oracle 11g` with same configuration, after switching to `Oracle 12c`  i got this error while `JOOQ Generator phase`, I cannot see the mapping of table or view columns POJOs are empty class ONLY constructor exists. After, checking generation log I find a lot of exception like below:

**Oracle version: 12c
JOOQ version: 3.12.3
JDK : 1.8**

**ACTUAL EXCEPTION**

`If you think this is a bug in jOOQ, please report it here: https://github.com/jOOQ/jOOQ/issues/new`

```sql
select 
  "SYS"."ALL_TAB_COLS"."DATA_TYPE_OWNER", 
  "SYS"."ALL_TAB_COLS"."DATA_TYPE", 
  decode(
    upper("SYS"."ALL_TAB_COLS"."DATA_TYPE"), 
    'CLOB', 
    0, 
    'BLOB', 
    0, 
    "SYS"."ALL_TAB_COLS"."CHAR_LENGTH"
  ) "char_length", 
  case
    when (
      "SYS"."ALL_TAB_COLS"."DATA_PRECISION" is null
      and "SYS"."ALL_TAB_COLS"."DATA_SCALE" = 0
    ) then 38
    else "SYS"."ALL_TAB_COLS"."DATA_PRECISION"
  end "DATA_PRECISION", 
  "SYS"."ALL_TAB_COLS"."DATA_SCALE", 
  "SYS"."ALL_TAB_COLS"."NULLABLE", 
  "SYS"."ALL_TAB_COLS"."DATA_DEFAULT", 
  "SYS"."ALL_TAB_COLS"."COLUMN_NAME", 
  "SYS"."ALL_TAB_COLS"."COLUMN_ID", 
  "SYS"."ALL_COL_COMMENTS"."COMMENTS", 
  nvl2("SYS"."ALL_TAB_IDENTITY_COLS"."COLUMN_NAME", 1, 0) "identity", 
  nvl2("SYS"."ALL_JSON_COLUMNS"."COLUMN_NAME", 1, 0) "json", 
  case "SYS"."ALL_JSON_COLUMNS"."DATA_TYPE"
    when 'BLOB' then 'JSONB'
    else 'JSON'
  end "jsonDataType"
from "SYS"."ALL_TAB_COLS"
  join "SYS"."ALL_COL_COMMENTS"
    on (
      "SYS"."ALL_TAB_COLS"."OWNER" = "SYS"."ALL_COL_COMMENTS"."OWNER"
      and "SYS"."ALL_TAB_COLS"."TABLE_NAME" = "SYS"."ALL_COL_COMMENTS"."TABLE_NAME"
      and "SYS"."ALL_TAB_COLS"."COLUMN_NAME" = "SYS"."ALL_COL_COMMENTS"."COLUMN_NAME"
    )
  left outer join "SYS"."ALL_TAB_IDENTITY_COLS"
    on (
      "SYS"."ALL_TAB_COLS"."OWNER" = "SYS"."ALL_TAB_IDENTITY_COLS"."OWNER"
      and "SYS"."ALL_TAB_COLS"."TABLE_NAME" = "SYS"."ALL_TAB_IDENTITY_COLS"."TABLE_NAME"
      and "SYS"."ALL_TAB_COLS"."COLUMN_NAME" = "SYS"."ALL_TAB_IDENTITY_COLS"."COLUMN_NAME"
    )
  left outer join "SYS"."ALL_JSON_COLUMNS"
    on (
      "SYS"."ALL_TAB_COLS"."OWNER" = "SYS"."ALL_JSON_COLUMNS"."OWNER"
      and "SYS"."ALL_TAB_COLS"."TABLE_NAME" = "SYS"."ALL_JSON_COLUMNS"."TABLE_NAME"
      and "SYS"."ALL_TAB_COLS"."COLUMN_NAME" = "SYS"."ALL_JSON_COLUMNS"."COLUMN_NAME"
    )
where (
  "SYS"."ALL_TAB_COLS"."OWNER" = '******'
  and "SYS"."ALL_TAB_COLS"."TABLE_NAME" = '******'
)
order by "SYS"."ALL_TAB_COLS"."COLUMN_ID"```
[ERROR] Error while initialising type
org.jooq.exception.DataAccessException: SQL [select "SYS"."ALL_TAB_COLS"."DATA_TYPE_OWNER", "SYS"."ALL_TAB_COLS"."DATA_TYPE", decode(upper("SYS"."ALL_TAB_COLS"."DATA_TYPE"), 'CLOB', 0, 'BLOB', 0, "SYS"."ALL_TAB_COLS"."CHAR_LENGTH") "char_length", case when ("SYS"."ALL_TAB_COLS"."DATA_PRECISION" is null and "SYS"."ALL_TAB_COLS"."DATA_SCALE" = 0) then 38 else "SYS"."ALL_TAB_COLS"."DATA_PRECISION" end "DATA_PRECISION", "SYS"."ALL_TAB_COLS"."DATA_SCALE", "SYS"."ALL_TAB_COLS"."NULLABLE", "SYS"."ALL_TAB_COLS"."DATA_DEFAULT", "SYS"."ALL_TAB_COLS"."COLUMN_NAME", "SYS"."ALL_TAB_COLS"."COLUMN_ID", "SYS"."ALL_COL_COMMENTS"."COMMENTS", nvl2("SYS"."ALL_TAB_IDENTITY_COLS"."COLUMN_NAME", 1, 0) "identity", nvl2("SYS"."ALL_JSON_COLUMNS"."COLUMN_NAME", 1, 0) "json", case "SYS"."ALL_JSON_COLUMNS"."DATA_TYPE" when 'BLOB' then 'JSONB' else 'JSON' end "jsonDataType" from "SYS"."ALL_TAB_COLS" join "SYS"."ALL_COL_COMMENTS" on ("SYS"."ALL_TAB_COLS"."OWNER" = "SYS"."ALL_COL_COMMENTS"."OWNER" and "SYS"."ALL_TAB_COLS"."TABLE_NAME" = "SYS"."ALL_COL_COMMENTS"."TABLE_NAME" and "SYS"."ALL_TAB_COLS"."COLUMN_NAME" = "SYS"."ALL_COL_COMMENTS"."COLUMN_NAME") left outer join "SYS"."ALL_TAB_IDENTITY_COLS" on ("SYS"."ALL_TAB_COLS"."OWNER" = "SYS"."ALL_TAB_IDENTITY_COLS"."OWNER" and "SYS"."ALL_TAB_COLS"."TABLE_NAME" = "SYS"."ALL_TAB_IDENTITY_COLS"."TABLE_NAME" and "SYS"."ALL_TAB_COLS"."COLUMN_NAME" = "SYS"."ALL_TAB_IDENTITY_COLS"."COLUMN_NAME") left outer join "SYS"."ALL_JSON_COLUMNS" on ("SYS"."ALL_TAB_COLS"."OWNER" = "SYS"."ALL_JSON_COLUMNS"."OWNER" and "SYS"."ALL_TAB_COLS"."TABLE_NAME" = "SYS"."ALL_JSON_COLUMNS"."TABLE_NAME" and "SYS"."ALL_TAB_COLS"."COLUMN_NAME" = "SYS"."ALL_JSON_COLUMNS"."COLUMN_NAME") where ("SYS"."ALL_TAB_COLS"."OWNER" = ? and "SYS"."ALL_TAB_COLS"."TABLE_NAME" = ?) order by "SYS"."ALL_TAB_COLS"."COLUMN_ID"]; ORA-00942: table or view does not exist
    at org.jooq_3.12.3.ORACLE.debug (Unknown Source)
    at org.jooq.impl.Tools.translate (Tools.java:2717)
    at org.jooq.impl.DefaultExecuteContext.sqlException (DefaultExecuteContext.java:755)
    at org.jooq.impl.AbstractQuery.execute (AbstractQuery.java:382)
    at org.jooq.impl.AbstractResultQuery.fetch (AbstractResultQuery.java:353)
    at org.jooq.impl.AbstractResultQuery.iterator (AbstractResultQuery.java:364)
    at org.jooq.impl.SelectImpl.iterator (SelectImpl.java:2703)
    at org.jooq.meta.oracle.OracleTableDefinition.getElements0 (OracleTableDefinition.java:105)
    at org.jooq.meta.AbstractElementContainerDefinition.getElements (AbstractElementContainerDefinition.java:84)
    at org.jooq.meta.AbstractTableDefinition.getColumns (AbstractTableDefinition.java:156)
    at org.jooq.meta.AbstractDatabase.syntheticPrimaryKeys (AbstractDatabase.java:2292)
    at org.jooq.meta.AbstractDatabase.getRelations0 (AbstractDatabase.java:2244)
    at org.jooq.meta.AbstractDatabase.getRelations (AbstractDatabase.java:1924)
    at org.jooq.meta.AbstractTableDefinition.getUniqueKeys (AbstractTableDefinition.java:97)
    at org.jooq.codegen.JavaGenerator.generateTable (JavaGenerator.java:4134)
    at org.jooq.codegen.JavaGenerator.generateTable (JavaGenerator.java:3818)
    at org.jooq.codegen.JavaGenerator.generateTables (JavaGenerator.java:3805)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:514)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:468)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:389)
    at org.jooq.codegen.GenerationTool.run0 (GenerationTool.java:800)
    at org.jooq.codegen.GenerationTool.run (GenerationTool.java:221)
    at org.jooq.codegen.GenerationTool.generate (GenerationTool.java:216)
    at org.jooq.codegen.maven.Plugin.execute (Plugin.java:198)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
Caused by: java.sql.SQLSyntaxErrorException: ORA-00942: table or view does not exist
    at oracle.jdbc.driver.T4CTTIoer.processError (T4CTTIoer.java:447)
    at oracle.jdbc.driver.T4CTTIoer.processError (T4CTTIoer.java:396)
    at oracle.jdbc.driver.T4C8Oall.processError (T4C8Oall.java:951)
    at oracle.jdbc.driver.T4CTTIfun.receive (T4CTTIfun.java:513)
    at oracle.jdbc.driver.T4CTTIfun.doRPC (T4CTTIfun.java:227)
    at oracle.jdbc.driver.T4C8Oall.doOALL (T4C8Oall.java:531)
    at oracle.jdbc.driver.T4CPreparedStatement.doOall8 (T4CPreparedStatement.java:208)
    at oracle.jdbc.driver.T4CPreparedStatement.executeForDescribe (T4CPreparedStatement.java:886)
    at oracle.jdbc.driver.OracleStatement.executeMaybeDescribe (OracleStatement.java:1175)
    at oracle.jdbc.driver.OracleStatement.doExecuteWithTimeout (OracleStatement.java:1296)
    at oracle.jdbc.driver.OraclePreparedStatement.executeInternal (OraclePreparedStatement.java:3613)
    at oracle.jdbc.driver.OraclePreparedStatement.execute (OraclePreparedStatement.java:3714)
    at oracle.jdbc.driver.OraclePreparedStatementWrapper.execute (OraclePreparedStatementWrapper.java:1378)
    at org.jooq.tools.jdbc.DefaultPreparedStatement.execute (DefaultPreparedStatement.java:209)
    at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet (Tools.java:3865)
    at org.jooq.impl.AbstractResultQuery.execute (AbstractResultQuery.java:294)
    at org.jooq.impl.AbstractQuery.execute (AbstractQuery.java:368)
    at org.jooq.impl.AbstractResultQuery.fetch (AbstractResultQuery.java:353)
    at org.jooq.impl.AbstractResultQuery.iterator (AbstractResultQuery.java:364)
    at org.jooq.impl.SelectImpl.iterator (SelectImpl.java:2703)
    at org.jooq.meta.oracle.OracleTableDefinition.getElements0 (OracleTableDefinition.java:105)
    at org.jooq.meta.AbstractElementContainerDefinition.getElements (AbstractElementContainerDefinition.java:84)
    at org.jooq.meta.AbstractTableDefinition.getColumns (AbstractTableDefinition.java:156)
    at org.jooq.meta.AbstractDatabase.syntheticPrimaryKeys (AbstractDatabase.java:2292)
    at org.jooq.meta.AbstractDatabase.getRelations0 (AbstractDatabase.java:2244)
    at org.jooq.meta.AbstractDatabase.getRelations (AbstractDatabase.java:1924)
    at org.jooq.meta.AbstractTableDefinition.getUniqueKeys (AbstractTableDefinition.java:97)
    at org.jooq.codegen.JavaGenerator.generateTable (JavaGenerator.java:4134)
    at org.jooq.codegen.JavaGenerator.generateTable (JavaGenerator.java:3818)
    at org.jooq.codegen.JavaGenerator.generateTables (JavaGenerator.java:3805)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:514)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:468)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:389)
    at org.jooq.codegen.GenerationTool.run0 (GenerationTool.java:800)
    at org.jooq.codegen.GenerationTool.run (GenerationTool.java:221)
    at org.jooq.codegen.GenerationTool.generate (GenerationTool.java:216)
    at org.jooq.codegen.maven.Plugin.execute (Plugin.java:198)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)

**Comment by lukaseder** [comment]
Thank you very much for your report. What exact Oracle version are you using?

```sql
select * from V$VERSION;
```

The main thing that has changed between jOOQ versions is that we now join `ALL_JSON_COLUMNS` and `ALL_TAB_IDENTITY_COLS`. These should be available, normally, in Oracle 12c, starting from 12.1:

- https://docs.oracle.com/database/121/REFRN/GUID-0FEC03C4-8F07-47C0-8FC0-A771EBF5D057.htm#REFRN20963
- https://docs.oracle.com/database/121/REFRN/GUID-2791B728-3670-4960-BEEE-EA750981C0DF.htm#REFRN20830

I'm assuming that `ALL_TAB_IDENTITY_COLS` is available in your case, because its availability is what we base our query upon, so it must be `ALL_JSON_COLUMNS`. Is that view not available in your installation, or are you missing grants?

**Comment by ahalim7** [comment]
Yes `SYS.ALL_JSON_COLUMNS` missing on mine, I will check grants with DB Administrator as well. 

**oracle version is : Oracle Database 12c Enterprise Edition Release 12.1.0.1.0 - 64bit Production **

**Comment by lukaseder** [comment]
I see `ALL_JSON_COLUMNS` was added in 12.1.0.2.0. We'll have to add an explicit check for that view's availability. Will fix this right away. A new patch release can be made this week to fix this for you.

What's the rationale to upgrade to an older version of Oracle 12c? It was, itself, quite full of bugs, e.g. related to the new adaptive query plan feature...

**Comment by ahalim7** [comment]
Highly Appreciate your support, and by the way, I have full grants on all sys tables. 
waiting for the release

**Comment by lukaseder** [comment]
> and by the way, I have full grants on all sys tables.

Indeed, if you're using 12.1.0.1.0, then this isn't a grants problem.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14.0. Backport scheduled for 3.13.5: https://github.com/jOOQ/jOOQ/issues/10496.

New 3.14.0-SNAPSHOT and 3.13.5-SNAPSHOT builds will be available soon (1-2 hours), containing the fix, from here: https://www.jooq.org/download/versions
